fast64_internal/mk64/f3d_course_parser.py

```diff
@@ -0,0 +1,338 @@
+import re
+import traceback
+from mathutils import Vector
+from ..f3d.f3d_gbi import F3D
+from ..f3d.f3d_material import RDPSettings
+from ..f3d.f3d_parser import F3DParsedCommands, ParsedMacro, math_eval, parseDLData, parseVertexData
+from ..f3d.f3d_writer import F3DVert
+from ..utility import PluginError, float_from_u16_str, gammaInverseValue, int_from_s16_str, readFile, unpackNormal
+
+
+def courseVertexFormatPatterns():
+    # position, uv, color/normal
+    return [
+        # decomp format
+        "\{\s*"
+        + "\{+([^,\}]*),([^,\}]*),([^,\}]*)\}\s*,\s*"
+        + "\{([^,\}]*),([^,\}]*)\}\s*,\s*"
+        + "\{MACRO_COLOR_FLAG\(([^,\}]*),([^,\}]*),([^,\}]*),([^,\}])*\),([^,\}]*)\}\s*"
+        + "\}",
+    ][0]
+
+
+def parseCourseVtx(path: str, f3d):
+    data = readFile(path)
+    pattern = courseVertexFormatPatterns()
+    vertexData = []
+    for values in re.findall(pattern, data, re.DOTALL):
+        values = [math_eval(g, f3d) for g in values]
+        vertexData.append(
+            F3DVert(
+                Vector(values[0:3]),
+                Vector(values[3:5]),
+                Vector(values[5:8]),
+                unpackNormal(values[8]),
+                values[9],
+            )
+        )
+    return vertexData
+
+
+def getVertexDataStart(vertexDataParam: str, f3d: F3D):
+    matchResult = re.search(r"\&?([A-Za-z0-9\_]*)\s*(\[([^\]]*)\])?\s*(\+(.*))?", vertexDataParam)
+    if matchResult is None:
+        raise PluginError("SPVertex param " + vertexDataParam + " is malformed.")
+
+    offset = 0
+    if matchResult.group(3):
+        offset += math_eval(matchResult.group(3), f3d)
+    if matchResult.group(5):
+        offset += math_eval(matchResult.group(5), f3d)
+
+    name = matchResult.group(1)
+
+    if matchResult.group(1).startswith("0x04"):
+        offset = (int(matchResult.group(1), 16) - 0x04000000) // 16
+        name = hex(0x04000000)
+    return name, offset
+
+
+def processCommands(self, dlData: str, dlName: str, dlCommands: "list[ParsedMacro]"):
```

Comment:
what is the difference with `fast64_internal/f3d/f3d_parser.py`'s `processCommands`? looks similar enough at a glance to want to refactor this into the generic one to avoid duplicate code

Reply:
the only difference is getVertexDataStart who are not the same function (even I use the same name)

Reply:
oh I see

what you want to do is subclass `F3DContext` then, to override its methods. for example `OOTF3DContext` does this

(`getVertexDataStart` is not a method of `F3DContext` currently but just move it in there)

Reply:
is just for me create a subclass for only modify one function is overkill but If you want to be like this ok